File: BuferRing/BufferRingArray.cpp
#include <iostream>
class BufferRingArray {
public:
	int start;
	int end;
	int maxSize;
	int count;
	int* buffer;

public:
	BufferRingArray(int Size) {
		start = 0;
		end = 0;
		maxSize = Size;
		count = 0;
		buffer = new int[maxSize];
	}
	bool isFull() {
		return count == maxSize;
	}
	bool isEmpty() {
		return count == 0;
	}
	void push_back(int data) {
		if (isFull()) {
			std::cout << "Buffer is full" << std::endl;
		}
		else {
			buffer[end] = data;
			end = (end + 1) % maxSize;
			count++;
		}
	}
	int pop_front() {
		if (isEmpty()) {
			std::cout << "Buffer is empty" << std::endl;
		}
		else {
			int data = buffer[start];
			start = (start + 1) % maxSize;
			count--;
			return data;
		}
	}
	int size() {
		return count;
	}
	~BufferRingArray() {
		delete[] buffer;
	}
};

int main() {
	BufferRingArray buffer(5);


	buffer.push_back(10);
	buffer.push_back(20);
	buffer.push_back(30);
	buffer.push_back(40);
	buffer.push_back(50);
	std::cout << "Delete: " << buffer.pop_front() << std::endl;
	std::cout << "Delete: " << buffer.pop_front() << std::endl;

	buffer.push_back(60);
	buffer.push_back(70);

	while (!buffer.isEmpty()) {
		std::cout << "Delete: " << buffer.pop_front() << std::endl;
	}

	buffer.pop_front();
}

File: BuferRing/BufferRingArray.py
class BufferRingArray:
    def __init__(self, size):
        self.start = 0
        self.end = 0
        self.max_size = size
        self.count = 0
        self.buffer = [0] * self.max_size

    def is_full(self):
        return self.count == self.max_size

    def is_empty(self):
        return self.count == 0

    def push_back(self, data):
        if self.is_full():
            print("Buffer is full")
        else:
            self.buffer[self.end] = data
            self.end = (self.end + 1) % self.max_size
            self.count += 1

    def pop_front(self):
        if self.is_empty():
            print("Buffer is empty")
        else:
            data = self.buffer[self.start]
            self.start = (self.start + 1) % self.max_size
            self.count -= 1
            return data

    def size(self):
        return self.count

def main():
    buffer = BufferRingArray(5)

    buffer.push_back(10)
    buffer.push_back(20)
    buffer.push_back(30)
    buffer.push_back(40)
    buffer.push_back(50)
    
    print("Delete:", buffer.pop_front())
    print("Delete:", buffer.pop_front())

    buffer.push_back(60)
    buffer.push_back(70)

    while not buffer.is_empty():
        print("Delete:", buffer.pop_front())

    buffer.pop_front()

if __name__ == "__main__":
    main()
